rd-net/Test.RdFramework/SocketProxy.cs

```diff
@@ -0,0 +1,194 @@
+using System;
+using System.IO;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading.Tasks;
+using JetBrains.Diagnostics;
+using JetBrains.Lifetimes;
+using JetBrains.Rd;
+using Test.Lifetimes;
+
+namespace Test.RdFramework
+{
+  public class SocketProxy
+  {
+    public readonly string Id;
+    private readonly Lifetime myLifetime;
+    private readonly int myServerPort;
+    private readonly ILog myLogger;
+
+    private int? myPort;
+
+    public int Port
+    {
+      get
+      {
+        if (myPort == null)
+        {
+          throw new InvalidOperationException("SocketProxy was not started");
+        }
+
+        return myPort.Value;
+      }
+    }
+
+    private const int DefaultBufferSize = 16370;
+    private readonly byte[] myServerToClientBuffer = new byte[DefaultBufferSize];
+    private readonly byte[] myClientToServerBuffer = new byte[DefaultBufferSize];
+
+    private readonly SequentialLifetimes myServerToClientLifetime;
+    private readonly SequentialLifetimes myClientToServerLifetime;
+    private TcpClient myProxyServer;
+    private TcpListener myProxyClient;
+
+    internal SocketProxy(string id, Lifetime lifetime, int serverPort)
+    {
+      Id = id;
+      myLifetime = lifetime;
+      myServerPort = serverPort;
+      myLogger = Log.GetLog<SocketProxy>().GetSublogger(id);
+
+      myServerToClientLifetime = new SequentialLifetimes(myLifetime).With(lifetimes => lifetimes.Next());
+      myClientToServerLifetime = new SequentialLifetimes(myLifetime).With(lifetimes => lifetimes.Next());
+
+      myLifetime.OnTermination(() =>
+      {
+        myPort = null;
+        
+        StopServerToClientMessaging();
+        StopClientToServerMessaging();
+
+        myProxyServer?.Close();
```

Comment:
Close(0) (see the other PR about netcore bugs)

Reply:
This is **TcpClient** and not **Socket**